File: app/src/main/java/uk/co/alt236/floatinginfo/ui/overlay/writers/MemoryTextWriter.java
/*
 * Copyright 2016 Alexandros Schillings
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package uk.co.alt236.floatinginfo.ui.overlay.writers;

import android.os.Build;

import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;
import uk.co.alt236.floatinginfo.data.access.generalinfo.inforeader.memory.MemoryData;
import uk.co.alt236.floatinginfo.data.prefs.EnabledInfoPrefs;
import uk.co.alt236.floatinginfo.util.StringBuilderHelper;
import uk.co.alt236.floatinginfo.util.Util;

/**
 *
 */
/*package*/ class MemoryTextWriter implements TextWriter<MemoryData> {
    private final PeakKeeper mPeakKeeper;
    private final EnabledInfoPrefs mEnabledInfo;

    public MemoryTextWriter(final EnabledInfoPrefs enabledInfoPrefs) {
        mPeakKeeper = new PeakKeeper();
        mEnabledInfo = enabledInfoPrefs;
    }

    @Override
    public void writeText(final MemoryData input, @NonNull final StringBuilderHelper sb) {
        if (input != null) {
            sb.appendBold("Current Process Memory Utilisation");

            if (input.getPid() < 0) {
                sb.appendBold("E: Invalid PID");

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    sb.appendBold("E: Since Nougat apps can no longer get the PID of other apps");
                }
            } else {
                sb.startKeyValueSection();
                constructMemoryLine(
                        sb,
                        "DalvikPrivateClean",
                        input.getDalvikPrivateClean());
                constructMemoryLine(
                        sb,
                        "DalvikPrivateDirty",
                        input.getDalvikPrivateDirty());
                constructMemoryLine(
                        sb,
                        "DalvikPss",
                        input.getDalvikPss());
                constructMemoryLine(
                        sb,
                        "DalvikSharedClean",
                        input.getDalvikSharedClean());
                constructMemoryLine(
                        sb,
                        "DalvikSharedDirty",
                        input.getDalvikSharedDirty());
                constructMemoryLine(
                        sb,
                        "DalvikSwappablePss",
                        input.getDalvikSwappablePss());
                constructMemoryLine(
                        sb,
                        "DalvikSwappedOut",
                        input.getDalvikSwappedOut());

                constructMemoryLine(
                        sb,
                        "NativePrivateClean",
                        input.getNativePrivateClean());
                constructMemoryLine(
                        sb,
                        "NativePrivateDirty",
                        input.getNativePrivateDirty());
                constructMemoryLine(
                        sb,
                        "NativePss",
                        input.getNativePss());
                constructMemoryLine(
                        sb,
                        "NativeSharedClean",
                        input.getNativeSharedClean());
                constructMemoryLine(
                        sb,
                        "NativeSharedDirty",
                        input.getNativeSharedDirty());
                constructMemoryLine(
                        sb,
                        "NativeSwappablePss",
                        input.getNativeSwappablePss());
                constructMemoryLine(
                        sb,
                        "NativeSwappedOut",
                        input.getNativeSwappedOut());

                constructMemoryLine(
                        sb,
                        "OtherPrivateClean",
                        input.getOtherPrivateClean());
                constructMemoryLine(
                        sb,
                        "OtherPrivateDirty",
                        input.getOtherPrivateDirty());
                constructMemoryLine(
                        sb,
                        "OtherPss",
                        input.getOtherPss());
                constructMemoryLine(
                        sb,
                        "OtherSharedClean",
                        input.getOtherSharedClean());
                constructMemoryLine(
                        sb,
                        "OtherSharedDirty",
                        input.getOtherSharedDirty());
                constructMemoryLine(
                        sb,
                        "OtherSwappablePss",
                        input.getOtherSwappablePss());
                constructMemoryLine(
                        sb,
                        "OtherSwappedOut",
                        input.getOtherSwappedOut());

                sb.endKeyValueSection();
            }
            sb.appendNewLine();
        }
    }

    @Override
    public void clear() {
        mPeakKeeper.clear();
    }

    private void constructMemoryLine(final StringBuilderHelper sb, final String key, final int value) {
        final boolean showZero = mEnabledInfo.showZeroMemoryItems();

        final boolean showItem;
        if (showZero) {
            showItem = value >= 0;
        } else if (value > 0) {
            showItem = true;
        } else {
            showItem = mPeakKeeper.getValue(key) > 0;
        }

        if (showItem) { // We assume negative values mean that something went
            // wrong with reflection.
            String valueStr = Util.getHumanReadableKiloByteCount(value, true);
            mPeakKeeper.update(key, value);
            final int peak = mPeakKeeper.getValue(key);
            if (peak > 0) {
                valueStr += " (Peak: " + Util.getHumanReadableKiloByteCount(peak, true) + ")";
            }

            sb.append(key, valueStr);
        }
    }


    private static class PeakKeeper {

        private final Object mLock = new Object();
        private final Map<String, Integer> mStore = new HashMap<String, Integer>();

        public void clear() {
            synchronized (mLock) {
                mStore.clear();
            }
        }

        public int getValue(final String key) {
            synchronized (mLock) {
                final Integer res = mStore.get(key);
                if (res == null) {
                    return 0;
                } else {
                    return res;
                }
            }
        }

        public void update(final String key, final int value) {
            synchronized (mLock) {
                final Integer stored = mStore.get(key);
                final int tmp;

                if (stored == null) {
                    tmp = -1;
                } else {
                    tmp = stored;
                }

                if (value > tmp) {
                    mStore.put(key, value);
                }
            }
        }
    }
}

File: overlay/src/main/java/uk/co/alt236/floatinginfo/overlay/writers/MemoryTextWriter.kt
/*
 * Copyright 2020 Alexandros Schillings
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.alt236.floatinginfo.overlay.writers

import android.os.Build
import uk.co.alt236.floatinginfo.common.data.model.MemoryData
import uk.co.alt236.floatinginfo.common.prefs.EnabledInfoPrefs
import uk.co.alt236.floatinginfo.common.string.HumanReadable.getHumanReadableKiloByteCount
import uk.co.alt236.floatinginfo.common.string.StringBuilderHelper
import java.util.*

/**
 *
 */
/*package*/
internal class MemoryTextWriter(private val enabledInfoPrefs: EnabledInfoPrefs) : TextWriter<MemoryData> {
    private val mPeakKeeper = PeakKeeper()

    override fun writeText(input: MemoryData?, sb: StringBuilderHelper) {
        if (input == null) {
            return
        }

        sb.appendBold("Current Process Memory Utilisation")
        if (input.pid < 0) {
            sb.appendBold("E: Invalid PID")
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                sb.appendBold("E: This is a Nougat+ restriction.")
            }
        } else {
            sb.startKeyValueSection()
            constructMemoryLine(
                    sb,
                    "DalvikPrivateClean",
                    input.dalvikPrivateClean)
            constructMemoryLine(
                    sb,
                    "DalvikPrivateDirty",
                    input.dalvikPrivateDirty)
            constructMemoryLine(
                    sb,
                    "DalvikPss",
                    input.dalvikPss)
            constructMemoryLine(
                    sb,
                    "DalvikSharedClean",
                    input.dalvikSharedClean)
            constructMemoryLine(
                    sb,
                    "DalvikSharedDirty",
                    input.dalvikSharedDirty)
            constructMemoryLine(
                    sb,
                    "DalvikSwappablePss",
                    input.dalvikSwappablePss)
            constructMemoryLine(
                    sb,
                    "DalvikSwappedOut",
                    input.dalvikSwappedOut)
            constructMemoryLine(
                    sb,
                    "NativePrivateClean",
                    input.nativePrivateClean)
            constructMemoryLine(
                    sb,
                    "NativePrivateDirty",
                    input.nativePrivateDirty)
            constructMemoryLine(
                    sb,
                    "NativePss",
                    input.nativePss)
            constructMemoryLine(
                    sb,
                    "NativeSharedClean",
                    input.nativeSharedClean)
            constructMemoryLine(
                    sb,
                    "NativeSharedDirty",
                    input.nativeSharedDirty)
            constructMemoryLine(
                    sb,
                    "NativeSwappablePss",
                    input.nativeSwappablePss)
            constructMemoryLine(
                    sb,
                    "NativeSwappedOut",
                    input.nativeSwappedOut)
            constructMemoryLine(
                    sb,
                    "OtherPrivateClean",
                    input.otherPrivateClean)
            constructMemoryLine(
                    sb,
                    "OtherPrivateDirty",
                    input.otherPrivateDirty)
            constructMemoryLine(
                    sb,
                    "OtherPss",
                    input.otherPss)
            constructMemoryLine(
                    sb,
                    "OtherSharedClean",
                    input.otherSharedClean)
            constructMemoryLine(
                    sb,
                    "OtherSharedDirty",
                    input.otherSharedDirty)
            constructMemoryLine(
                    sb,
                    "OtherSwappablePss",
                    input.otherSwappablePss)
            constructMemoryLine(
                    sb,
                    "OtherSwappedOut",
                    input.otherSwappedOut)
            sb.endKeyValueSection()
        }
        sb.appendNewLine()
    }

    override fun clear() {
        mPeakKeeper.clear()
    }

    private fun constructMemoryLine(sb: StringBuilderHelper, key: String, value: Int) {
        val showZero = enabledInfoPrefs.showZeroMemoryItems()
        val showItem: Boolean

        showItem = when {
            showZero -> value >= 0
            value > 0 -> true
            else -> mPeakKeeper.getValue(key) > 0
        }

        if (!showItem) { // We assume negative values mean that something went wrong with reflection.
            return
        }

        var valueStr = getHumanReadableKiloByteCount(value.toLong(), true)
        mPeakKeeper.update(key, value)
        val peak = mPeakKeeper.getValue(key)
        if (peak > 0) {
            valueStr += " (Peak: " + getHumanReadableKiloByteCount(peak.toLong(), true) + ")"
        }
        sb.append(key, valueStr)
    }

    private class PeakKeeper {
        private val mLock = Any()
        private val mStore: MutableMap<String, Int> = HashMap()
        fun clear() {
            synchronized(mLock) { mStore.clear() }
        }

        fun getValue(key: String?): Int {
            synchronized(mLock) {
                val res = mStore[key]
                return res ?: 0
            }
        }

        fun update(key: String, value: Int) {
            synchronized(mLock) {
                val stored = mStore[key]
                val tmp: Int
                tmp = stored ?: -1
                if (value > tmp) {
                    mStore[key] = value
                }
            }
        }
    }
}